signac/__main__.py

```diff
@@ -3,379 +3,192 @@
-    def hashpw(pw):
-        if scheme is None:
-            return pw
-        else:
-            return get_crypt_context().encrypt(
-                pw, scheme=scheme)
-    hostcfg = config['hosts'][hostname]
-    hostcfg['password'] = get_credentials(hostcfg)
-    db_auth = get_database(
-        hostcfg.get('db_auth', 'admin'),
-        hostname=hostname, config=config)
-    if new_pw is None:
-        new_pw = _prompt_for_new_password()
-    pwhash = hashpw(new_pw)
-    db_auth.add_user(hostcfg['username'], pwhash)
-    return pwhash
-
-
 def _read_index(project, fn_index=None):
     if fn_index is not None:
         _print_err("Reading index from file '{}'...".format(fn_index))
         fd = open(fn_index)
         return (json.loads(l) for l in fd)
 
 
-def _open_job_by_id(project, job_id):
-    "Attempt to open a job by id and provide user feedback on error."
-    try:
-        return project.open_job(id=job_id)
-    except KeyError:
-        close_matches = difflib.get_close_matches(
-            job_id, [jid[:len(job_id)] for jid in project.find_job_ids()])
-        msg = "Did not find job corresponding to id '{}'.".format(job_id)
-        if len(close_matches) == 1:
-            msg += " Did you mean '{}'?".format(close_matches[0])
-        elif len(close_matches) > 1:
-            msg += " Did you mean any of [{}]?".format('|'.join(close_matches))
-        raise KeyError(msg)
-    except LookupError:
-        n = project.min_len_unique_id()
-        raise LookupError("Multiple matches for abbreviated id '{}'. "
-                          "Use at least {} characters for guaranteed "
-                          "unique ids.".format(job_id, n))
-
+def transform_option(opt):
+    if len(opt) < 1:
+        return None
+    if opt[0] is None:
+        return list()
+    return list(opt)
 
-def find_with_filter_or_none(args):
-    if args.job_id or args.filter or args.doc_filter:
-        return find_with_filter(args)
 
-
-def find_with_filter(args):
-    if getattr(args, 'job_id', None):
-        if args.filter or args.doc_filter:
+def find_with_filter(**kwargs):
+    if getattr(kwargs, 'job_id', None):
+        if kwargs['filter'] or kwargs['doc_filter']:
             raise ValueError("Can't provide both 'job-id' and filter arguments!")
         else:
-            return args.job_id
+            return kwargs['job_id']
 
     project = get_project()
-    if hasattr(args, 'index'):
-        index = _read_index(project, args.index)
+    if hasattr(kwargs, 'index'):
+        index = _read_index(project, kwargs['index'])
     else:
         index = None
 
-    f = parse_filter_arg(args.filter)
-    df = parse_filter_arg(args.doc_filter)
+    f = parse_filter_arg(kwargs['filter'])
+    df = parse_filter_arg(kwargs['doc_filter'])
     return get_project().find_job_ids(index=index, filter=f, doc_filter=df)
 
 
-def main_project(args):
+class MultipleOptionalArgument(click.Option):
```

Comment:
I'm not sure if I understand the purpose of this class. @vishav1771 Can you offer some explanation for why this exists? It seemed to me that "Multi Value Options" or "Multiple Options" would suffice, but I'm not sure if I understand the purpose. https://click.palletsprojects.com/en/7.x/options/#multi-value-options

Reply:
I tried to match the original command. As click does not support `nargs=*, +, ?`, so I need to do some monkey patching. This class customize the `add_to_parser` func and set the arguments according to need.
[https://stackoverflow.com/questions/48391777/nargs-equivalent-for-options-in-click](https://stackoverflow.com/questions/48391777/nargs-equivalent-for-options-in-click)

The "Multiple Option" was the other option. But I found a few issues:

- For `--sp`: I need to divide this command into two commands `--all-sp`, `--sp` where `--all-sp` will work as `--sp`(previous api) with no argument and `--sp` will work as `--sp`(previous api).
- For `--doc-filter`: command `--doc-filter a 0` will be written as `--doc-filter a --doc-filter 0` which do not make sense. And if I go for  tuple of two values then I was not able to create `--doc-filter a`   {"a": {"$exists": true}}

Kindly provide the feedback.

Reply:
These are the exact issues I anticipated in attempting an exact replication of the current interface and I'm glad that we are having this discussion.

The click framework intentionally does not support options with variadic arguments, because it creates syntactical ambiguities. This makes sense to me and I'd prefer to deprecated these instead of trying to force such behavior into the new framework.

However, there is [development on enabling options with default arguments](https://github.com/pallets/click/pull/1508), which should make it much easier to replicate the existing interface.

@vishav1771 Could you make a suggestion as to how we could revise the find interface to not rely on variadic options, but assuming that flag options are possible?

Then we can identify how we can use the kind of work-arounds you introduced here to enable backwards compatible behavior with deprecation warnings.

Reply:
> However, there is [development on enabling options with default arguments](https://github.com/pallets/click/pull/1508), which should make it much easier to replicate the existing interface.

Taking this into account, I think `--sp` should not be changed. But we need to change `--doc-filter`. 
We can implement `--doc-filter key:value` with setting multiple to True. So `--doc-filter a 0 b 1 c` would look like `--doc-filter a:0 --doc-filter b:1 --doc-filter c`. In my opinion, arguments to `find` should be also be changed (it can be implemented as it is) in a similar way. (`find a 0 b 1 c` to `find a:0 b:1 c`)

Reply:
> Taking this into account, I think `--sp` should not be changed. 

Are you sure about that? Since `--sp` is currently accepting an arbitrary number of arguments and I don't see that supported even with the patch.

> But we need to change `--doc-filter`.
> We can implement `--doc-filter key:value` with setting multiple to True. So `--doc-filter a 0 b 1 c` would look like `--doc-filter a:0 --doc-filter b:1 --doc-filter c`. 

Afaik, it is possible to specify the exact number arguments an option accepts, so while it's true that users would need to use multiple options to create a filter, changing the notation to the `key:value` syntax would not strictly be necessary, right?

> In my opinion, arguments to `find` should be also be changed (it can be implemented as it is) in a similar way. (`find a 0 b 1 c` to `find a:0 b:1 c`)

I think we should be much more careful with considerations concerning changing the API for the arguments. Users are very much used to these now and it would present a huge break in backwards compatibility.

I propose we keep the discussion going and I will also experiment with this a little bit myself.

Reply:
>Afaik, it is possible to specify the exact number arguments an option accepts, so while it's true that users would need to use multiple options to create a filter, changing the notation to the key:value syntax would not strictly be necessary, right?

It is not strictly necessary. But i think `--doc-filter a:0` is much more intuitive than `--doc-filter a --doc-filter 0` 

> I think we should be much more careful with considerations concerning changing the API for the arguments. Users are very much used to these now and it would present a huge break in backwards compatibility. 

Makes sense. (I just wanted to keep both the filters similar.)

Reply:
>Are you sure about that? Since --sp is currently accepting an arbitrary number of arguments and I don't see that supported even with the patch.

I think that should be achievable by setting multiple to True. I will try and let you know.

Reply:
I tried and was able to implement `sp` option as  `--sp`, `--sp=a --sp=b`.

Reply:
Yes, that's fine!